linux/dm-sworndisk/src/bio.rs

```diff
@@ -0,0 +1,413 @@
+
+/// A wrapper for kernel's struct `bio`.
+pub struct Bio {
+    inner: NonNull<bindings::bio>,
+}
+
+// SAFETY: we get a reference to kernel's `bio` by `bio_get`, so it won't disappear
+// and can be safely transferred across threads.
+unsafe impl Send for Bio {}
+
+impl Bio {
+    /// Constructs a `Bio` from raw pointer.
+    ///
+    /// # Safety
+    ///
+    /// User must provide a valid pointer to the kernel's `bio`.
+    pub unsafe fn from_raw(ptr: *mut bindings::bio) -> Self {
+        // Get a reference to a `bio`, so it won't disappear. And `bio_put` is
+        // called when it's been dropped.
+        bindings::bio_get(ptr);
+        Self {
+            inner: NonNull::new_unchecked(ptr),
+        }
+    }
+
+    /// Allocates a bio for the target `BlockDevice`.
+    ///
+    /// The bio operation is specified by `op`, targeting a given block region,
+    /// while its I/O buffer is configured by `BioVec`.
+    pub fn alloc(
+        bdev: &BlockDevice,
+        op: BioOp,
+        region: Range<usize>,
+        io_buffer: &[BioVec],
+    ) -> Result<Self> {
+        // SAFETY: no safety requirements on this FFI call.
+        let bio = unsafe {
+            bindings::bio_alloc_bioset(
+                bdev.as_ptr(),
+                io_buffer.len() as _,
+                op as _,
+                GFP_NOIO,
+                &mut bindings::fs_bio_set,
+            )
+        };
+        if bio.is_null() {
+            return Err(ENOMEM);
+        }
+
+        let mut buf_len = 0usize;
+        for bio_vec in io_buffer {
+            // SAFETY: the `bio` is non-null and valid (checked above), and `bio_vecs`
+            // contains valid pages.
+            unsafe {
+                let page = bindings::virt_to_page(bio_vec.ptr as _);
+                bindings::__bio_add_page(bio, page, bio_vec.len as _, bio_vec.offset as _);
+                buf_len += bio_vec.len;
+            }
+        }
+
+        let start_sector = region.start * BLOCK_SECTORS;
+        let nr_bytes = region.len() * BLOCK_SIZE;
+        if nr_bytes != buf_len {
+            // SAFETY: the `bio` is non-null and valid. And the 'bio_alloc_bioset'
+            // method sets the reference count, so we call `bio_put` to drop it here.
+            unsafe { bindings::bio_put(bio) };
+            return Err(EINVAL);
+        }
+        // SAFETY: the `bio` is non-null and valid.
+        unsafe {
+            (*bio).bi_iter.bi_sector = start_sector as _;
+            (*bio).bi_iter.bi_size = nr_bytes as _;
+        }
+
+        // SAFETY: the `bio` is non-null and valid, and the 'bio_alloc_bioset' method
+        // sets the reference count, so we don't need to call `bio_get` here.
+        Ok(Self {
+            inner: unsafe { NonNull::new_unchecked(bio) },
+        })
+    }
+
+    // Clones a bio that shares the original bio's I/O buffer.
+    pub fn alloc_clone(&self) -> Result<Self> {
```

Comment:
Since `bio_alloc_clone` function is an existing C concept, choosing a name as close as reasonably possible to the C side should avoid confusion and improve readability when switching back and forth between the C and Rust sides.